### ISSUE TITLE
[KF1.2] GCP ComputeDisks missing required spec fields

### ISSUE THREAD

**Issue by jcastagneri** [kubeflow/gcp-blueprints#194]
# Version Information
KF1.2; installation following the [canonical instructions](https://www.kubeflow.org/docs/gke/deploy/deploy-cli/) from a clean GCP project.

# Summary
Below, I debug an issue with the cnrm compute disk manifests. During application of these resources, I received 4xx errors saying that my resource spec was missing needed fields. I was able to fix it by adding more configuration to the upstream resources to specify `replicaZones`, and by bumping up the size. I'm curious if this is a recent break in GCP's API for computeDisks or if I have something set up in an abnormal way.

# Leadup to the error
I was applying a configured KF1.2 cluster with `make apply`. I expected to successfully deploy the cluster.

# Error
When applying the kubeflow cluster using the GCP blueprint, I ran across this error:
```
77 Error from server (Forbidden): error when creating ".build/kubeflow-apps/v1_persistentvolume_minio-pv.yaml": persistentvolumes "minio-pv" is forbidden: error querying GCE PD volume amp-kubeflow-storage-artifact-store: disk is not found
778 Error from server (Forbidden): error when creating ".build/kubeflow-apps/v1_persistentvolume_mysql-pv.yaml": persistentvolumes "mysql-pv" is forbidden: error querying GCE PD volume amp-kubeflow-storage-metadata-store: disk is not found
```

# Related Issue
I found [this](https://github.com/kubeflow/manifests/issues/1294) issue where a missing kustomize path was added. *This is not the source of my problem; I verified that my resource names are overriding correctly as seen above (amp-kubeflow is my project name)*.

# Debugging steps
I searched for the persistent disk these volumes are looking to bind to. They had failed to create:
```
> kubectl --context=kubeflow-mgmt describe computedisk.compute.cnrm.cloud.google.com/amp-kubeflow-storage-artifact-store

Events:
  Type     Reason        Age                   From                    Message
  ----     ------        ----                  ----                    -------
  Normal   Updating      3m37s (x19 over 25m)  computedisk-controller  Update in progress
  Warning  UpdateFailed  3m37s (x19 over 25m)  computedisk-controller  Update call failed: error applying desired state: summary: Error creating RegionDisk: googleapi: Error 400: Invalid value for field 'resource.replicaZones[0]': ''. Exactly 2 replica zones must be set for regional disks., invalid, detail:
```

Indeed, I found the resource definition in `amp-kubeflow/upstream/manifests/gcp/v2/cnrm/pipelines/disk.yaml`:
```
apiVersion: compute.cnrm.cloud.google.com/v1beta1
kind: ComputeDisk
metadata:
  name: amp-kubeflow-storage-metadata-store # {"$kpt-set":"name-storage-metadata-store"}
spec:
  location: us-east1 # {"$kpt-set":"location"}
  size: 20
---
apiVersion: compute.cnrm.cloud.google.com/v1beta1
kind: ComputeDisk
metadata:
  name: amp-kubeflow-storage-artifact-store # {"$kpt-set":"name-storage-artifact-store"}
spec:
  location: us-east1 # {"$kpt-set":"location"}
  size: 200
```

If I change these resources to specify `replicaZones`, they apply successfully. Changed versions below:
```
apiVersion: compute.cnrm.cloud.google.com/v1beta1
kind: ComputeDisk
metadata:
  name: amp-kubeflow-storage-metadata-store # {"$kpt-set":"name-storage-metadata-store"}
spec:
  location: us-east1 # {"$kpt-set":"location"}
  size: 200
  replicaZones:
    - us-east1-b
    - us-east1-c
---
apiVersion: compute.cnrm.cloud.google.com/v1beta1
kind: ComputeDisk
metadata:
  name: amp-kubeflow-storage-artifact-store # {"$kpt-set":"name-storage-artifact-store"}
spec:
  location: us-east1 # {"$kpt-set":"location"}
    size: 200
  replicaZones:
    - us-east1-b
    - us-east1-c
```

I don't see setters for zones available in the package, so I just hardcoded them into my version of the upstream code. 

**Note** I also had to increase the size of one of the disks to 200GB; the API would give a different 400 error saying the minimum disk size is 200GB otherwise.

As stated above, these changes seem to fix the issue.

# Questions
Is this something I did wrong? I'm fairly new to `kpt`, and while this appears to be API drift on GCP's side, I am curious to see if I have anything set up in a weird way that would've caused this. My install is in a fresh GCP project and completely isolated, with nothing custom done to it, so I can't think of anything I did weird (one never can, though).

If this is an issue requiring a change to the blueprint, is this best handled by adding a setter for zone(s)?

**Comment by Bobgy** [comment]
@jcastagneri thank you for the report! I'm glad you found a workaround for yourself.

I suspect this is a subset of https://github.com/kubeflow/gcp-blueprints/issues/6.
The setup wasn't made work well out-of-box for regional clusters. Your issues seem to be one of that.

**Comment by Bobgy** [comment]
I think https://github.com/kubeflow/pipelines/issues/4356 might be a better long term solution, use GCS & Cloud SQL as the backing stores instead of persistence disk. This makes operational costs lower.

**Comment by jcastagneri** [comment]
Thanks @Bobgy. I read through your linked issues; for my vote, managed storage makes pragmatic sense. It would be easier for migration, a bit less frightening for data loss, and more economical. I wish PDs were easier to manage in GKE, or that there were better tools to work with them, but managed storage seems to be winning right now. I think you're making the right call here if you're changing the defaults (looks like its planned for 1.3).

**Comment by Bobgy** [comment]
Let's mark this as closed, will be resolved in kubeflow/pipelines#4356.
